compiler/rustc_const_eval/src/interpret/validity.rs

```diff
@@ -707,20 +681,56 @@ impl<'rt, 'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> ValidityVisitor<'rt, 'mir, '
     fn in_mutable_memory(&self, op: &OpTy<'tcx, M::Provenance>) -> bool {
         if let Some(mplace) = op.as_mplace_or_imm().left() {
             if let Some(alloc_id) = mplace.ptr().provenance.and_then(|p| p.get_alloc_id()) {
-                let mutability = match self.ecx.tcx.global_alloc(alloc_id) {
-                    GlobalAlloc::Static(_) => {
-                        self.ecx.memory.alloc_map.get(alloc_id).unwrap().1.mutability
-                    }
-                    GlobalAlloc::Memory(alloc) => alloc.inner().mutability,
-                    _ => span_bug!(self.ecx.tcx.span, "not a memory allocation"),
-                };
-                return mutability == Mutability::Mut;
+                return mutability(self.ecx, alloc_id).0.is_mut();
             }
         }
         false
     }
 }
 
+/// Returns whether the allocation is mutable, and whether it's actually a static.
+/// For "root" statics we look at the type to account for interior
+/// mutability; for nested statics we have no type and directly use the annotated mutability.
+fn mutability<'mir, 'tcx: 'mir>(
+    ecx: &InterpCx<'mir, 'tcx, impl Machine<'mir, 'tcx>>,
+    alloc_id: AllocId,
+) -> (Mutability, Option<(DefId, bool)>) {
+    // Let's see what kind of memory this points to.
+    // We're not using `try_global_alloc` since dangling pointers have already been handled.
+    match ecx.tcx.global_alloc(alloc_id) {
+        GlobalAlloc::Static(did) => {
+            let DefKind::Static { mutability, nested } = ecx.tcx.def_kind(did) else { bug!() };
+            let mutability = if nested {
+                mutability
+            } else {
+                let mutability = match mutability {
+                    Mutability::Not
+                        if !ecx
+                            .tcx
+                            .type_of(did)
+                            .no_bound_vars()
+                            .expect("statics should not have generic parameters")
+                            .is_freeze(*ecx.tcx, ty::ParamEnv::reveal_all()) =>
+                    {
+                        Mutability::Mut
+                    }
+                    _ => mutability,
+                };
+                if let Some((_, alloc)) = ecx.memory.alloc_map.get(alloc_id) {
+                    assert_eq!(alloc.mutability, mutability);
```

Comment:
Can this assertion be moved outside the `if nested { ... } else { ... }`?

Reply:
There will never be an entry for nested statics. They have already been interned. I'll assert that there is no such entry.

Reply:
Ah I had not even noticed that this specifically checks the `alloc_map`. Why that? Is it possible to use `get_alloc_mutability` instead and do this sanity check for all statics, nested or otherwise?

The only thing that has not been interned here is the root alloc of a static, right?

Reply:
We can't use that as it actually fetches the allocation, and that will cause cycle errors for recursive statics

```rust
static S: *const u8 = &S as *const *const u8 as *const u8;
```

has a different `AllocId` for `&S` than it has for the root allocation. The root allocation's `AllocId` will never be leaked, but we still need it for evaluating the static itself. While this case can be detected, mutually recursive statics are impossible to detect locally.

Reply:
I am confused -- I thought we use the same AllocId for the return place as for the static?

Reply:
Hmm right. No idea where the cycle errors are coming from then. I just tested it and they still occur.

Reply:
If you called `get_alloc_raw` or something like that, that's still querying the static so would lead to a cycle if it is the current static.

Reply:
Ah this is an ICE fix as well it seems, nice. :D